io.c

```diff
@@ -3377,69 +3397,77 @@ io_write_nonblock(rb_execution_context_t *ec, VALUE io, VALUE str, VALUE ex)
- *  Reads _length_ bytes from the I/O stream.
+ *  Reads (in binary mode) up to +maxlen+ bytes from the stream.
  *
- *  _length_ must be a non-negative integer or +nil+.
+ *  If +maxlen+ is +nil+, reads the entire stream and returns the content as a string;
+ *  encoding conversion is applied, if applicable:
  *
- *  If _length_ is a positive integer, +read+ tries to read
- *  _length_ bytes without any conversion (binary mode).
- *  It returns +nil+ if an EOF is encountered before anything can be read.
- *  Fewer than _length_ bytes are returned if an EOF is encountered during
- *  the read.
- *  In the case of an integer _length_, the resulting string is always
- *  in ASCII-8BIT encoding.
+ *    f = File.new('t.txt')
+ *    f.read # => "This is line one.\nThis is the second line.\nThis is the third line.\n"
+ *    f.eof  # => true
+ *    f.read # => ""
  *
- *  If _length_ is omitted or is +nil+, it reads until EOF
- *  and the encoding conversion is applied, if applicable.
- *  A string is returned even if EOF is encountered before any data is read.
+ *  If +maxlen+ is zero, returns an empty string:
  *
- *  If _length_ is zero, it returns an empty string (<code>""</code>).
+ *    f = File.new('t.txt')
+ *    f.read(0) # => ""
  *
- *  If the optional _outbuf_ argument is present,
- *  it must reference a String, which will receive the data.
- *  The _outbuf_ will contain only the received data after the method call
- *  even if it is not empty at the beginning.
+ *  If +maxlen+ is positive and the stream has unread data,
+ *  returns a new string with +maxlen+ bytes from the stream, if available,
+ *  or all available bytes otherwise;
+ *  the returned string has ASCII-8BIT encoding:
  *
- *  When this method is called at end of file, it returns +nil+
- *  or <code>""</code>, depending on _length_:
- *  +read+, <code>read(nil)</code>, and <code>read(0)</code> return
- *  <code>""</code>,
- *  <code>read(<i>positive_integer</i>)</code> returns +nil+.
+ *    f = File.new('t.txt')
+ *    f.read(30) # => "This is line one.\r\nThis is the"
+ *    f.read(30) # => " second line.\r\nThis is the thi"
+ *    f.read(30) # => "rd line.\r\n"
  *
- *     f = File.new("testfile")
- *     f.read(16)   #=> "This is line one"
+ *  Returns +nil+ if +maxlen+ is positive and the stream has no unread data:
  *
- *     # read whole file
- *     open("file") do |f|
- *       data = f.read   # This returns a string even if the file is empty.
- *       # ...
- *     end
+ *    # Stream was emptied by reads above.
+ *    f.read(30) # => nil
  *
- *     # iterate over fixed length records
- *     open("fixed-record-file") do |f|
- *       while record = f.read(256)
- *         # ...
- *       end
- *     end
+ *  With integer argument +maxlen+ and string argument +out_string+ given,
+ *  returns +out_string+ with +maxlen+ bytes from the stream, if available,
+ *  or all available bytes otherwise;
+ *  if any bytes are read, they replace the content of +out_string+:
```

Comment:
`out_string` is replaced by the empty string if no bytes are read:

```ruby
s = nil; File.open('ruby.c'){|f| f.read; p f.read(0, s = "x")}; s # => ''
s = nil; File.open('ruby.c'){|f| f.read; p f.read(1, s = "x")}; s # => ''
s = nil; File.open('ruby.c'){|f| f.read; p f.read(nil, s = "x")}; s # => ''
```

Reply:
Fixed.